# Portfolio Optimization

### Libraries

In [3]:
# Core Libraries 
import pandas as pd

import numpy as np

# Data Sets
import yfinance as yf

# Date Opitimization
import datetime

We will pull stock price data using the Yahoo Finance library. You can easily install the library using pip in a terminal command line

In [6]:
start_date = "2019-01-01"

We should pull stocks from a few different industries, so we’ll gather price data in healthcare, tech, retail and finance.

Let’s wrap this logic in a function that we can easily reuse since we will be pulling several stocks:

In [9]:
def get_stock(ticker):
    data = yf.download(ticker)
    
    data = data[data.index >= start_date]
    
    data[f'{ticker}'] = data["Close"]
    
    data = data[[f'{ticker}']] 
    
    print(data.head())
    
    return data 

In [10]:
get_stock("MRNA")

[*********************100%%**********************]  1 of 1 completed

                 MRNA
Date                 
2019-01-02  15.330000
2019-01-03  15.500000
2019-01-04  16.959999
2019-01-07  16.270000
2019-01-08  16.950001



C:\Users\Batuhan\AppData\Local\Temp\ipykernel_153036\3781743562.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data[f'{ticker}'] = data["Close"]


MRNA
Date                  
2019-01-02   15.330000
2019-01-03   15.500000
2019-01-04   16.959999
2019-01-07   16.270000
2019-01-08   16.950001
...                ...
2024-03-04   96.000000
2024-03-05   95.559998
2024-03-06   98.099998
2024-03-07   99.480003
2024-03-08  103.029999

[1305 rows x 1 columns]

Let’s define another function that takes a list of stocks and generate a single data frame of stock prices for each stock:

In [11]:
from functools import reduce

def combine_stocks(tickers):
    data_frames = []
    for i in tickers:
        data_frames.append(get_stock(i))
        
    df_merged = reduce(lambda  left,right: pd.merge(left,right,on=['Date'], how='outer'), data_frames)
    print(df_merged.head())
    return df_merged


stocks = ["MRNA", "PFE", "JNJ"]
combine_stocks(stocks)

[*********************100%%**********************]  1 of 1 completed
C:\Users\Batuhan\AppData\Local\Temp\ipykernel_153036\3781743562.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data[f'{ticker}'] = data["Close"]


                 MRNA
Date                 
2019-01-02  15.330000
2019-01-03  15.500000
2019-01-04  16.959999
2019-01-07  16.270000
2019-01-08  16.950001


[*********************100%%**********************]  1 of 1 completed
C:\Users\Batuhan\AppData\Local\Temp\ipykernel_153036\3781743562.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data[f'{ticker}'] = data["Close"]


                  PFE
Date                 
2019-01-02  41.034157
2019-01-03  39.886147
2019-01-04  40.796963
2019-01-07  41.015179
2019-01-08  41.204933


[*********************100%%**********************]  1 of 1 completed

                   JNJ
Date                  
2019-01-02  127.750000
2019-01-03  125.720001
2019-01-04  127.830002
2019-01-07  127.010002
2019-01-08  129.960007
                 MRNA        PFE         JNJ
Date                                        
2019-01-02  15.330000  41.034157  127.750000
2019-01-03  15.500000  39.886147  125.720001
2019-01-04  16.959999  40.796963  127.830002
2019-01-07  16.270000  41.015179  127.010002
2019-01-08  16.950001  41.204933  129.960007



C:\Users\Batuhan\AppData\Local\Temp\ipykernel_153036\3781743562.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data[f'{ticker}'] = data["Close"]


MRNA        PFE         JNJ
Date                                         
2019-01-02   15.330000  41.034157  127.750000
2019-01-03   15.500000  39.886147  125.720001
2019-01-04   16.959999  40.796963  127.830002
2019-01-07   16.270000  41.015179  127.010002
2019-01-08   16.950001  41.204933  129.960007
...                ...        ...         ...
2024-03-04   96.000000  25.889999  159.839996
2024-03-05   95.559998  26.080000  159.970001
2024-03-06   98.099998  27.190001  159.339996
2024-03-07   99.480003  26.790001  158.869995
2024-03-08  103.029999  27.219999  159.520004

[1305 rows x 3 columns]

Healthcare: Moderna (MRNA), Pfizer (PFE), Johnson & Johnson (JNJ)

Tech: Google (GOOGL),  Microsoft (MSFT), Apple (AAPL)

Retail: Costco (COST), Walmart (WMT),  Kroger Co (KR)

Finance: JPMorgan Chase & Co (JPM), Bank of America (BAC), HSBC Holding (HSBC)

In [19]:
stocks = ["MRNA", "PFE", "JNJ", "GOOGL", 
          "MSFT", "AAPL", "COST", "WMT", "KR", "JPM", 
          "BAC", "HSBC"]
portfolio = combine_stocks(stocks)

[*********************100%%**********************]  1 of 1 completed
C:\Users\Batuhan\AppData\Local\Temp\ipykernel_153036\3781743562.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data[f'{ticker}'] = data["Close"]


                 MRNA
Date                 
2019-01-02  15.330000
2019-01-03  15.500000
2019-01-04  16.959999
2019-01-07  16.270000
2019-01-08  16.950001


[*********************100%%**********************]  1 of 1 completed
C:\Users\Batuhan\AppData\Local\Temp\ipykernel_153036\3781743562.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data[f'{ticker}'] = data["Close"]


                  PFE
Date                 
2019-01-02  41.034157
2019-01-03  39.886147
2019-01-04  40.796963
2019-01-07  41.015179
2019-01-08  41.204933


[*********************100%%**********************]  1 of 1 completed
C:\Users\Batuhan\AppData\Local\Temp\ipykernel_153036\3781743562.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data[f'{ticker}'] = data["Close"]
[*********************100%%**********************]  1 of 1 completed
C:\Users\Batuhan\AppData\Local\Temp\ipykernel_153036\3781743562.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data[f'{ticker}'] = data["Close"]


                   JNJ
Date                  
2019-01-02  127.750000
2019-01-03  125.720001
2019-01-04  127.830002
2019-01-07  127.010002
2019-01-08  129.960007
                GOOGL
Date                 
2019-01-02  52.734001
2019-01-03  51.273499
2019-01-04  53.903500
2019-01-07  53.796001
2019-01-08  54.268501


[*********************100%%**********************]  1 of 1 completed
C:\Users\Batuhan\AppData\Local\Temp\ipykernel_153036\3781743562.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data[f'{ticker}'] = data["Close"]


                  MSFT
Date                  
2019-01-02  101.120003
2019-01-03   97.400002
2019-01-04  101.930000
2019-01-07  102.059998
2019-01-08  102.800003


[*********************100%%**********************]  1 of 1 completed
C:\Users\Batuhan\AppData\Local\Temp\ipykernel_153036\3781743562.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data[f'{ticker}'] = data["Close"]


                 AAPL
Date                 
2019-01-02  39.480000
2019-01-03  35.547501
2019-01-04  37.064999
2019-01-07  36.982498
2019-01-08  37.687500


[*********************100%%**********************]  1 of 1 completed
C:\Users\Batuhan\AppData\Local\Temp\ipykernel_153036\3781743562.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data[f'{ticker}'] = data["Close"]


                  COST
Date                  
2019-01-02  204.759995
2019-01-03  200.419998
2019-01-04  206.240005
2019-01-07  207.000000
2019-01-08  208.550003


[*********************100%%**********************]  1 of 1 completed
C:\Users\Batuhan\AppData\Local\Temp\ipykernel_153036\3781743562.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data[f'{ticker}'] = data["Close"]


                  WMT
Date                 
2019-01-02  31.113333
2019-01-03  30.953333
2019-01-04  31.146667
2019-01-07  31.513332
2019-01-08  31.733334


[*********************100%%**********************]  1 of 1 completed
C:\Users\Batuhan\AppData\Local\Temp\ipykernel_153036\3781743562.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data[f'{ticker}'] = data["Close"]


                   KR
Date                 
2019-01-02  27.299999
2019-01-03  27.350000
2019-01-04  27.660000
2019-01-07  27.920000
2019-01-08  28.459999


[*********************100%%**********************]  1 of 1 completed
C:\Users\Batuhan\AppData\Local\Temp\ipykernel_153036\3781743562.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data[f'{ticker}'] = data["Close"]


                   JPM
Date                  
2019-01-02   99.309998
2019-01-03   97.110001
2019-01-04  100.690002
2019-01-07  100.760002
2019-01-08  100.570000


[*********************100%%**********************]  1 of 1 completed
C:\Users\Batuhan\AppData\Local\Temp\ipykernel_153036\3781743562.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data[f'{ticker}'] = data["Close"]


                  BAC
Date                 
2019-01-02  24.959999
2019-01-03  24.559999
2019-01-04  25.580000
2019-01-07  25.559999
2019-01-08  25.510000


[*********************100%%**********************]  1 of 1 completed

                 HSBC
Date                 
2019-01-02  40.880001
2019-01-03  40.430000
2019-01-04  41.610001
2019-01-07  41.049999
2019-01-08  41.160000
                 MRNA        PFE         JNJ      GOOGL        MSFT  \
Date                                                                  
2019-01-02  15.330000  41.034157  127.750000  52.734001  101.120003   
2019-01-03  15.500000  39.886147  125.720001  51.273499   97.400002   
2019-01-04  16.959999  40.796963  127.830002  53.903500  101.930000   
2019-01-07  16.270000  41.015179  127.010002  53.796001  102.059998   
2019-01-08  16.950001  41.204933  129.960007  54.268501  102.800003   

                 AAPL        COST        WMT         KR         JPM  \
Date                                                                  
2019-01-02  39.480000  204.759995  31.113333  27.299999   99.309998   
2019-01-03  35.547501  200.419998  30.953333  27.350000   97.110001   
2019-01-04  37.064999  206.240005  31.146667  27.660000  100.690


C:\Users\Batuhan\AppData\Local\Temp\ipykernel_153036\3781743562.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data[f'{ticker}'] = data["Close"]


We now have a single dataframe of returns for our stocks. Let’s write this dataframe to a csv so we can easily read in the data without repeatedly having to pull it using the Pandas-Datareader. 


In [20]:
portfolio.to_csv("portfolio.csv", index=False)

#Now, let’s read in our csv:

portfolio = pd.read_csv("portfolio.csv")

## Mean Variance Optimization

**pip install PyPortfolioOpt** Please upload this libary inisde of Terminal

Now, let’s calculate the covariance matrix and store the calculated returns in variables S and mu, respectively:

In [21]:
from pypfopt.expected_returns import mean_historical_return
from pypfopt.risk_models import CovarianceShrinkage


mu = mean_historical_return(portfolio)
S = CovarianceShrinkage(portfolio).ledoit_wolf()

Next, let’s import the EfficientFrontier module and calculate the weights. Here, we will use the max Sharpe statistic. The Sharpe ratio is the ratio between returns and risk. The lower the risk and the higher the returns, the higher the Sharpe ratio. The algorithm looks for the maximum Sharpe ratio, which translates to the portfolio with the highest return and lowest risk. Ultimately, the higher the Sharpe ratio, the better the performance of the portfolio. 

In [22]:
from pypfopt.efficient_frontier import EfficientFrontier

ef = EfficientFrontier(mu, S)
weights = ef.max_sharpe()

cleaned_weights = ef.clean_weights()
print(dict(cleaned_weights))

{'MRNA': 0.09243, 'PFE': 0.0, 'JNJ': 0.0, 'GOOGL': 0.0, 'MSFT': 0.14485, 'AAPL': 0.19979, 'COST': 0.49321, 'WMT': 0.0, 'KR': 0.06973, 'JPM': 0.0, 'BAC': 0.0, 'HSBC': 0.0}


We can also display portfolio performance:

In [23]:
ef.portfolio_performance(verbose=True)

Expected annual return: 29.8%
Annual volatility: 22.4%
Sharpe Ratio: 1.24


(0.29811983686203747, 0.22367683372741642, 1.243400276315463)

In [24]:
portfolio

MRNA        PFE         JNJ       GOOGL        MSFT        AAPL  \
0      15.330000  41.034157  127.750000   52.734001  101.120003   39.480000   
1      15.500000  39.886147  125.720001   51.273499   97.400002   35.547501   
2      16.959999  40.796963  127.830002   53.903500  101.930000   37.064999   
3      16.270000  41.015179  127.010002   53.796001  102.059998   36.982498   
4      16.950001  41.204933  129.960007   54.268501  102.800003   37.687500   
...          ...        ...         ...         ...         ...         ...   
1300   96.000000  25.889999  159.839996  133.350006  414.920013  175.100006   
1301   95.559998  26.080000  159.970001  132.669998  402.649994  170.119995   
1302   98.099998  27.190001  159.339996  131.399994  402.089996  169.119995   
1303   99.480003  26.790001  158.869995  134.380005  409.140015  169.000000   
1304  103.029999  27.219999  159.520004  135.410004  406.220001  170.729996   

            COST        WMT         KR         JPM        BAC       HSBC  
0     204.759995  31.113333  27.299999   99.309998  24.959999  40.880001  
1     200.419998  30.953333  27.350000   97.110001  24.559999  40.430000  
2     206.240005  31.146667  27.660000  100.690002  25.580000  41.610001  
3     207.000000  31.513332  27.920000  100.760002  25.559999  41.049999  
4     208.550003  31.733334  28.459999  100.570000  25.510000  41.160000  
...          ...        ...        ...         ...        ...        ...  
1300  759.179993  59.299999  49.369999  186.679993  35.150002  38.779999  
1301  759.950012  60.040001  49.480000  188.550003  35.389999  38.930000  
1302  773.239990  60.570000  50.490002  189.529999  35.419998  39.230000  
1303  785.590027  60.360001  55.480000  187.869995  35.630001  37.880001  
1304  725.559998  60.119999  55.970001  188.220001  35.599998  37.439999  

[1305 rows x 12 columns]

Finally, let’s convert the weights into actual allocations values (i.e., how many of each stock to buy). For our allocation, let’s consider an investment amount of $100,000:

In [25]:
from pypfopt.discrete_allocation import DiscreteAllocation, get_latest_prices

latest_prices = get_latest_prices(portfolio)

da = DiscreteAllocation(weights, latest_prices, total_portfolio_value=100000)

allocation, leftover = da.greedy_portfolio()
print("Discrete allocation:", allocation)
print("Funds remaining: ${:.2f}".format(leftover))

Discrete allocation: {'COST': 68, 'AAPL': 117, 'MSFT': 35, 'MRNA': 90, 'KR': 125}
Funds remaining: $199.86


Our algorithm says we should invest in 90 shares of MRNA, 35 shares of MSFT, 117 shares of AAPL and 125 shares of KR. 

We see that our portfolio performs with an expected annual return of 125 percent. This performance is due to the rapid growth of Moderna during the pandemic. Further, the Sharpe ratio value of 5.02 indicates that the portfolio optimization algorithm performs well with our current data. Of course, this return is inflated and is not likely to hold up in the future. 

Mean variance optimization doesn’t perform very well since it makes many simplifying assumptions, such as returns being normally distributed and the need for an invertible covariance matrix. Fortunately, methods like HRP and mCVAR address these limitations. 

## Hierarchical Risk Parity (HRP)

The HRP method works by finding subclusters of similar assets based on returns and constructing a hierarchy from these clusters to generate weights for each asset. 

In [27]:
from pypfopt import HRPOpt

# We need to calculate the returns:
returns = portfolio.pct_change().dropna()

#Then run the optimization algorithm to get the weights:
hrp = HRPOpt(returns)
hrp_weights = hrp.optimize()

#We can now print the performance of the portfolio and the weights:
hrp.portfolio_performance(verbose=True)
print(dict(hrp_weights))

Expected annual return: 14.6%
Annual volatility: 17.0%
Sharpe Ratio: 0.74
{'AAPL': 0.04327661197033801, 'BAC': 0.058823069826983945, 'COST': 0.0819700179904097, 'GOOGL': 0.04384106797270626, 'HSBC': 0.06134639415514409, 'JNJ': 0.214609631461112, 'JPM': 0.05175059384646395, 'KR': 0.112081871114903, 'MRNA': 0.02016684498594857, 'MSFT': 0.08147984476209509, 'PFE': 0.13456937454359905, 'WMT': 0.09608467737029636}


C:\Users\Batuhan\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\pypfopt\hierarchical_portfolio.py:138: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '[0.59452471 0.59452471 0.59452471 0.59452471 0.59452471 0.59452471]' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  w[first_cluster] *= alpha  # weight 1


We see that we have an expected annual return of 14.6 percent, which is significantly less than the inflated 125 percent we achieved with mean variance optimization. We also see a diminished Sharpe ratio of 0.74.  This result is much more reasonable and more likely to hold up in the future since HRP is not as sensitive to outliers as mean variance optimization is. 

In [28]:
da_hrp = DiscreteAllocation(hrp_weights, latest_prices, total_portfolio_value=100000)

allocation, leftover = da_hrp.greedy_portfolio()
print("Discrete allocation (HRP):", allocation)
print("Funds remaining (HRP): ${:.2f}".format(leftover))

Discrete allocation (HRP): {'JNJ': 134, 'PFE': 494, 'KR': 200, 'WMT': 160, 'COST': 11, 'MSFT': 20, 'HSBC': 164, 'BAC': 165, 'JPM': 28, 'GOOGL': 33, 'AAPL': 26, 'MRNA': 20}
Funds remaining (HRP): $6.45


## Mean Conditional Value at Risk (mCVAR)

The mCVAR is another popular alternative to mean variance optimization. It works by measuring the worst-case scenarios for each asset in the portfolio, which is represented here by losing the most money. The worst-case loss for each asset is then used to calculate weights to be used for allocation for each asset. 

In [29]:
from pypfopt.efficient_frontier import EfficientCVaR

#Calculate the weights and get the performance:
S = portfolio.cov()
ef_cvar = EfficientCVaR(mu, S)
cvar_weights = ef_cvar.min_cvar()

cleaned_weights = ef_cvar.clean_weights()
print(dict(cleaned_weights))

{'MRNA': 0.0, 'PFE': 0.0, 'JNJ': 0.0, 'GOOGL': 0.0, 'MSFT': 0.0, 'AAPL': 0.0, 'COST': 1.0, 'WMT': 0.0, 'KR': 0.0, 'JPM': 0.0, 'BAC': 0.0, 'HSBC': 0.0}


C:\Users\Batuhan\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\cvxpy\reductions\solvers\solving_chain.py:336: FutureWarning: 
    Your problem is being solved with the ECOS solver by default. Starting in 
    CVXPY 1.5.0, Clarabel will be used as the default solver instead. To continue 
    using ECOS, specify the ECOS solver explicitly using the ``solver=cp.ECOS`` 
    argument to the ``problem.solve`` method.
    
  warnings.warn(ECOS_DEPRECATION_MSG, FutureWarning)


In [30]:
# Next, get the discrete allocation:
da_cvar = DiscreteAllocation(cvar_weights, latest_prices, total_portfolio_value=100000)

allocation, leftover = da_cvar.greedy_portfolio()
print("Discrete allocation (CVAR):", allocation)
print("Funds remaining (CVAR): ${:.2f}".format(leftover))

Discrete allocation (CVAR): {'COST': 137, 'HSBC': 1, 'JPM': 1, 'GOOGL': 1, 'KR': 1, 'BAC': 1, 'WMT': 1, 'PFE': 1}
Funds remaining (CVAR): $58.30


Although we only considered healthcare, tech, retail and finance, the methods we discussed can easily be modified to consider additional industries. For example, maybe you are more interested in constructing a portfolio of companies in the energy, real estate and materials industry. An example of this sort of portfolio could be made up of stocks such as Exxonmobil (XOM), DuPont (DD), and American Tower (AMT). I encourage you to play around with different sectors in constructing your portfolio. 

What we discussed provides a solid foundation for those interested in portfolio optimization methods in Python. Having a knowledge of both the methods and the tools available for portfolio optimization can allow quants and data scientists to run quicker experiments for optimizing investment portfolio.